modules/gcloud/src/main/java/org/testcontainers/containers/BigQueryEmulatorContainer.java

```diff
@@ -25,15 +46,220 @@ public BigQueryEmulatorContainer(String image) {
     public BigQueryEmulatorContainer(DockerImageName dockerImageName) {
         super(dockerImageName);
         dockerImageName.assertCompatibleWith(DEFAULT_IMAGE_NAME);
-        addExposedPorts(HTTP_PORT, GRPC_PORT);
-        withCommand("--project", PROJECT_ID);
+    }
+
+    @Override
+    protected void configure() {
+        StringBuilder command = new StringBuilder();
+        command.append("bigquery-emulator --project=" + projectId + " --port=" + httpPort + " --grpc-port=" + grpcPort);
+        dataset.ifPresent(it -> command.append(" --dataset=" + it));
+        logLevel.ifPresent(it -> command.append(" --log-level=" + it));
+        logFormat.ifPresent(it -> command.append(" --log-format=" + it));
+        database.ifPresent(it -> command.append(" --database=" + it));
+        dataFromYaml.ifPresent(it -> command.append(" --data-from-yaml=" + it));
+        withCommand(command.toString());
+        addFixedExposedPort(httpPort, httpPort);
+        addFixedExposedPort(grpcPort, grpcPort);
```

Comment:
anti-pattern. See https://www.atomicjar.com/2023/11/testcontainers-best-practices/

Reply:
The ports are not fixed. They are chosen dynamically by default. See also
`private int httpPort = TestSocketUtils.findAvailableTcpPort();`

As we need to ensure that the emulator running in the container must have the same port as exposed to the Docker host, we can't rely on Docker assigning a free random port (as done by default in testcontainers as far as I can judge from my investigation). I therefore selected a random free port directly in Java.